### PR TITLE
[MetaSchedule] Enable anchor-block tuning

### PR DESCRIPTION
**Pull request by masahi**
Note: Most diffs are from test cases, which are bloated due to many TVMScript modules.

Building on the notion of "module equality" introduced in https://github.com/apache/tvm/pull/13050, I'm adding a new variant of module equality based on the "anchor blocks". I defined the anchor block in https://github.com/apache/tvm/pull/13194. 

Currently, MS does tuning at the level of subgraphs. So for example in resnet where there are `conv2d -> add -> relu` and `conv2d -> add -> add -> relu` subgraphs, the two subgraphs are treated as distinct tuning tasks even if the anchor block conv2d workloads are identical. The new module equality will identify them as equal, so it reduces the number of tuning tasks if shorter tuning time is preferred over subgraph-level performance. Currently there is no dedicated API for anchor-block tuning - passing `module_equality="anchor-block"` to task extraction and `tune_relay` will enable it. 

This is particularly effective for int8 models, since each conv2d / dense (anchor) op is quantized slightly differently, ending up many similar but not identical elemwise ops fused after the anchor blocks. On the int8 resnet50 model I tested, it reduced the number of conv2d tuning tasks from 36 to 23.   

The interesting question is the performance difference between full subgraph-level tuning and anchor-block tuning. To experiment this, I tested the int8 resnet50 mentioned above, where anchor-based tuning makes the most difference in terms of the # of extracted tasks. The results are summarized below. 

`num_iter_per_task = 32`, and `max_iters_per_task = 128` in all cases.
| Target | Anchor tuning | Subgraph tuning |
| ---- | ---- | ---- |
| x86 VNNI | 4.45 | 4.58 |
| Hexagon | 58.4 | 58.1 |
|CUDA tensor core (batch size 16) | 6.0 | 6.7 (TODO: Try again)|

From the model + target combinations I tested, I didn't see much perf difference beyond those from natural tuning flaky-ness. The tensorcore result is a bit weird and needs more investigation, but I found that tuning this model using int8 tensor core auto tensorization is incredibly slow currently (for example, getting the 6.7 result took 12 hours). There is also the correctness issue discussed in https://github.com/apache/tvm/issues/13204. So I haven't done more experiments on tensorcore.

### Applying the trace tuned on an anchor block to the target block 
The tricky problem that this work address is the application of a trace, which is tuned on a "representative" anchor subgraph, to the target mod which has different post blocks. Note that, in the resnet example where there are `conv2d -> add -> relu` and `conv2d -> add -> add -> relu` subgraphs, we tune the "smaller" `conv2d -> add -> relu` subgraph, not just the pure anchor block `conv2d`. 

So while applying a trace tuned on just `conv2d` to `conv2d -> add` would be trivial (the existing `Trace::ApplyToSchedule` would just work), in practice we would be applying a trace tuned on `conv2d -> add` to `conv2d -> subtract`, for example. My proposed solution is implemented in `src/meta_schedule/trace_apply.cc` and it is tested extensively in `test_meta_schedule_trace_apply.py`.

@junrushao @zxybazh @vinx13 @tkonolige

### CODE REVIEW

**Review by junrushao — APPROVED** [review]
This is just amazing!! Thanks for the PR!

**Review by Hzfengsy — APPROVED** [review]
LGTM. Thanks @masahi for the great work!

File: src/meta_schedule/trace_apply.cc [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+  if (sch->Get(b1)->name_hint == sch->Get(b2)->name_hint) {
+    return true;
+  }
+  for (auto prod : sch->GetProducers(b2)) {
+    if (IsAncestor(b1, prod, sch)) return true;
+  }
+  return false;
+}
+
+void InlinePostBlocks(Schedule sch, Trace anchor_trace, Target target) {
```
Comment (zxybazh) [inline src/meta_schedule/trace_apply.cc:47]: Can you please add some function description here?

File: src/tir/schedule/primitive/cache_read_write.cc [hunk 2]
```diff
@@ -297,7 +297,12 @@ bool CalculateAffineFlag(const ScheduleState& self, const StmtSRef& block_sref)
  * \param stage The stage to be inserted
  * \return A SeqStmt, the result after insertion
  */
-SeqStmt InsertCacheStage(const Stmt& stmt, int pos, const Stmt& stage) {
+Stmt InsertCacheStage(const Stmt& stmt, int pos, const Stmt& stage) {
```
Comment (zxybazh) [inline src/tir/schedule/primitive/cache_read_write.cc:300]: Is this change related?

File: src/meta_schedule/trace_apply.cc [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+  for (auto name : block_names_orig) {
+    if (block_names_now.count(name) && name != "root" && !is_scheduled(name)) {
+      unscheduled_blocks.push_back(sch->GetBlock(name));
+    }
+  }
+
+  return unscheduled_blocks;
+}
+
+void ScheduleUsingAnchorTrace(Schedule sch, const Trace& anchor_trace, const tvm::Target& target) {
```
Comment (zxybazh) [inline src/meta_schedule/trace_apply.cc:196]: Ditto

**Review by zxybazh — APPROVED** [review]
Apart from a few nits looks pretty good to me, very delicate design and comprehensive tests! Thanks @masahi!

File: src/meta_schedule/trace_apply.cc (hunk 3 above)
Comment (masahi) [reply src/meta_schedule/trace_apply.cc:196]: This is documented in the header.

File: src/tir/schedule/primitive/cache_read_write.cc (hunk 2 above)
Comment (masahi) [reply src/tir/schedule/primitive/cache_read_write.cc:300]: It is not related but I found this bug while testing this branch. It happens on CPU when `link-params = True` is used and `cache_write` is applied during tuning.

Comment (masahi) [reply src/tir/schedule/primitive/cache_read_write.cc:300]: Added a test case for this fix. For some reason I cannot reproduce this bug without anchor tuning. So I'm including it here.
